src/composables/secureStorageRef.ts

```diff
@@ -0,0 +1,146 @@
+import { Ref, ref, watch } from 'vue';
+import { validateMnemonic } from '@aeternity/bip39';
+
+import type { StorageKey } from '@/types';
+import { decrypt, encrypt, watchUntilTruthy } from '@/utils';
+import { IS_MOBILE_APP, STORAGE_KEYS } from '@/constants';
+import { WalletStorage } from '@/lib/WalletStorage';
+import { useAccounts } from './accounts';
+import { type ICreateStorageRefOptions, useStorageRef } from './storageRef';
+
+/**
+ * Create a secure storage ref that encrypts and decrypts the value using the password.
+ * Is a wrapper around the useStorageRef composable.
+ */
+export function useSecureStorageRef<T = string | object | any[]>(
+  initialState: T,
+  storageKey: StorageKey,
+  options: Omit<
+    ICreateStorageRefOptions<T>,
+    'serializer' | 'isSecure' | 'onBackgroundSync'
+  > = {},
+) {
+  const isWriting = ref(false);
+
+  async function getDecryptedValue(encryptedValue: any) {
+    if (IS_MOBILE_APP || !encryptedValue) {
+      return encryptedValue;
+    }
+    watchUntilTruthy(() => !isWriting.value);
+    const { passwordKey } = useAccounts();
+    await watchUntilTruthy(passwordKey);
+    if (passwordKey.value) {
+      try {
+        const decryptedValue = await decrypt(passwordKey.value, encryptedValue);
+        try {
+          return JSON.parse(decryptedValue);
+        } catch (e) {
+          return decryptedValue;
+        }
+      } catch (e) {
+        throw new Error('Failed to decrypt the value');
+      }
+    }
+    throw new Error('Failed to read the value, probably the password is not set');
+  }
+
+  async function getEncryptedValue(val: any): Promise<T> {
+    if (IS_MOBILE_APP || !val) {
+      return val;
+    }
+    isWriting.value = true;
```

Comment:
Why this is called `isWriting`? At the beginning I thought it relates to moment when some sort of state is saved, but it is used here when the state is actually read so I think the name is misleading.

Reply:
I'm removing this flag as I don't think it is providing any value anymore